src/replica/duplication/replica_duplicator_manager.cpp

```diff
@@ -191,5 +198,30 @@ replica_duplicator_manager::get_dup_states() const
     return ret;
 }
 
+std::map<dupid_t, std::string> replica_duplicator_manager::get_progress_messages() const
+{
+    std::map<dupid_t, std::string> dup_messages;
+
+    zauto_lock l(_lock);
+
+    std::transform(_duplications.begin(),
+                   _duplications.end(),
+                   std::inserter(dup_messages, dup_messages.end()),
+                   [](const decltype(_duplications)::value_type &dup) {
+                       const auto &progress = dup.second->progress();
+                       return std::make_pair(
+                           dup.first,
+                           fmt::format("dupid={}, remote_cluster_name={}, remote_app_name={}, "
+                                       "confirmed_decree={}, persisted_decree={}",
+                                       dup.first,
+                                       dup.second->remote_cluster_name(),
```

Comment:
How about add a stringfy function for replica_duplicator, then we don't need to access its members one by one here?